external-crates/move/crates/move-compiler/src/parser/syntax.rs

```diff
@@ -32,25 +36,68 @@ impl<'env, 'lexer, 'input> Context<'env, 'lexer, 'input> {
         tokens: &'lexer mut Lexer<'input>,
         package_name: Option<Symbol>,
     ) -> Self {
+        let stop_set = TokenSet::from([Tok::EOF]);
         Self {
             package_name,
             env,
             tokens,
+            stop_set,
+        }
+    }
+
+    /// Checks if the current token is a member of the stop set.
+    fn at_stop_set(&self) -> bool {
+        self.tokens.at_set(&self.stop_set)
+    }
+
+    /// Advances tokens until reaching an element of the stop set, recording diagnostics along the
+    /// way (including the first optional one passed as an argument).
+    fn advance_until_stop_set(&mut self, diag_opt: Option<Diagnostic>) {
+        if let Some(diag) = diag_opt {
+            self.add_diag(diag);
+        }
+        while !self.at_stop_set() {
+            if let Err(err) = self.tokens.advance() {
+                self.add_diag(*err);
+            }
         }
     }
 
     fn at_end(&self, prev: Loc) -> bool {
         prev.end() as usize == self.tokens.start_loc()
     }
+
+    /// Advances token and records a resulting diagnostic (if any).
+    fn advance(&mut self) {
+        if let Err(diag) = self.tokens.advance() {
+            self.add_diag(*diag);
+        }
+    }
+
+    fn add_diag(&mut self, diag: Diagnostic) {
+        if diag.info().category() == Category::Syntax as u8
+            && diag.info().code() == Syntax::UnexpectedToken as u8
```

Comment:
Or maybe just not let the error get to this point. I'm not entirely sure where these are coming from, so hard to say.

Reply:
> Or maybe just not let the error get to this point. I'm not entirely sure where these are coming from, so hard to say.

This is meant to (conservatively) catch unexpected EOF `syntax::unexpected_token_error` but `UnexpectedToken` error is also constructed by-hand  and sometimes by-hand, so it might be tough to guarantee that EOF is always reported via `UnexpectedEOF` rather than `UnexpectedToken`, particularly if someone adds more by-hand cases in the future.

I can certainly still try introducing the new `UnexpectedEOF` code and update existing uses of `UnexpectedToken` to try and use it in cases when EOF is encountered, but it does feel a bit ad-hoc, so please let me know if I should go that route.

Reply:
> it does feel a bit ad-hoc

To me this check is already grossly ad-hoc, and I would rather just report unexpected EOF. But if we want to keep it, I would rather have it be well structured than checking for a particular message

Reply:
I don't love this check either, and while it affects quite a few tests, it may not be that much of a problem in practice. I removed it so that you can see if you can tolerate the result. Happy to work on making it more structured if you do not.